zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsServiceImpl.java

```diff
@@ -0,0 +1,278 @@
+            return AsyncTaskResult.taskResult();
+        }
+
+        if (getSupportedLocales(targetProjectSlug, targetVersionSlug).isEmpty()) {
+            log.error("No locales enabled in target version of {} [{}]",
+                    targetProjectSlug, targetVersionSlug);
+            return AsyncTaskResult.taskResult();
+        }
+
+        Optional<MergeTranslationsTaskHandle> taskHandleOpt =
+                Optional.fromNullable(handle);
+
+        if (taskHandleOpt.isPresent()) {
+            prepareMergeTranslationsHandle(sourceVersion, targetVersion,
+                    taskHandleOpt.get());
+        }
+
+        Stopwatch overallStopwatch = Stopwatch.createStarted();
+        log.info("merge translations start: from {} to {}", sourceProjectSlug
+                + ":" + sourceVersionSlug, targetProjectSlug + ":"
+                + targetVersionSlug);
+
+        int startCount = 0;
+        int totalCount = getTotalMatchCount(sourceVersion.getId(),
+                targetVersion.getId());
+
+        List<HLocale> supportedLocales = getSupportedLocales(targetVersion
+                .getProject().getSlug(), targetVersion.getSlug());
+
+        while (startCount < totalCount) {
+            int processedCount =
+                    mergeTranslationBatch(sourceVersion, targetVersion,
+                            supportedLocales, useNewerTranslation, startCount,
+                            TRANSLATION_BATCH_SIZE);
+            if (taskHandleOpt.isPresent()) {
+                taskHandleOpt.get().increaseProgress(processedCount);
+            }
+
+            startCount += TRANSLATION_BATCH_SIZE;
+            textFlowDAO.clear();
+        }
+        versionStateCacheImpl.clearVersionStatsCache(targetVersion.getId());
+        log.info("merge translation end: from {} to {}, {}", sourceProjectSlug
+                + ":" + sourceVersionSlug, targetProjectSlug + ":"
+                + targetVersionSlug, overallStopwatch);
+
+        return AsyncTaskResult.taskResult();
+    }
+
+    protected int mergeTranslationBatch(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion, List<HLocale> supportedLocales,
+            boolean useNewerTranslation, int offset, int batchSize) {
+        try {
+            return new MergeTranslationsWork(sourceVersion.getId(),
+                    targetVersion.getId(), offset, batchSize,
+                    useNewerTranslation, supportedLocales,
+                    textFlowDAO, translationStateCacheImpl)
+                    .workInTransaction();
+        } catch (Exception e) {
+            log.warn("exception during copy text flow target", e);
+            return 0;
+        }
+    }
+
+    /**
+     * Check if sourceVersion or targetVersion has source document.
+     *
+     * @param sourceVersion
+     * @param targetVersion
+     */
+    private boolean isVersionsEmpty(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion) {
```

Comment:
I'm not saying we definitely shouldn't do it but we need to separate the lists and have a better selection method.

I was thinking a drop down with tabs and the initial activated tab is 'my projects' and the other tab is all projects.

I think initially this would work with just the ones you maintain. If you really want to copy another one they can add you as a maintainer.

Do we have any use cases for copying from a project you don't maintain?

> On 16 Mar 2015, at 6:46 am, Alex Eng <notifications@github.com> wrote:
> 
> In zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsServiceImpl.java:
> 
> > +                    textFlowDAO, translationStateCacheImpl)
> > +                    .workInTransaction();
> > +        } catch (Exception e) {
> > +            log.warn("exception during copy text flow target", e);
> > +            return 0;
> > +        }
> > +    }
> > +
> > +    /**
> > +     * Check if sourceVersion or targetVersion has source document.
> > +     *
> > +     * @param sourceVersion
> > +     * @param targetVersion
> > +     */
> > +    private boolean isVersionsEmpty(HProjectIteration sourceVersion,
> > +            HProjectIteration targetVersion) {
> How about user who wish to copy from other projects?
> 
> —
> Reply to this email directly or view it on GitHub.
>